SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverDebugContext.cpp

```diff
@@ -0,0 +1,404 @@
+void USpatialNetDriverDebugContext::OnDebugComponentAuthLost(Worker_EntityId EntityId)
+{
+	for (auto Iterator = ActorDebugInfo.CreateIterator(); Iterator; ++Iterator)
+	{
+		if (Iterator->Value.Entity == EntityId)
+		{
+			Iterator.RemoveCurrent();
+			break;
+		}
+	}
+}
+
+void USpatialNetDriverDebugContext::AddEntityToWatch(Worker_EntityId Entity)
+{
+	bool bAlreadyWatchingEntity = false;
+	CachedInterestSet.Add(Entity, &bAlreadyWatchingEntity);
+	bNeedToUpdateInterest |= !bAlreadyWatchingEntity;
+}
+
+void USpatialNetDriverDebugContext::RemoveEntityToWatch(Worker_EntityId Entity)
+{
+	if (CachedInterestSet.Remove(Entity) > 0)
+	{
+		bNeedToUpdateInterest = true;
+	}
+}
+
+void USpatialNetDriverDebugContext::AddInterestOnTag(FName Tag)
+{
+	bool bAlreadyInSet = false;
+	SemanticInterest.Add(Tag, &bAlreadyInSet);
+
+	if (!bAlreadyInSet)
+	{
+		TArray<Worker_EntityId_Key> EntityIds;
+		NetDriver->StaticComponentView->GetEntityIds(EntityIds);
+
+		for (auto Entity : EntityIds)
+		{
+			if (SpatialGDK::DebugComponent* DbgComp = NetDriver->StaticComponentView->GetComponentData<SpatialGDK::DebugComponent>(Entity))
+			{
+				if (DbgComp->ActorTags.Contains(Tag))
+				{
+					AddEntityToWatch(Entity);
+				}
+			}
+		}
+
+		for (const auto& Item : ActorDebugInfo)
+		{
+			if (!Item.Value.bAdded || Item.Value.bUpdated)
+			{
+				if (Item.Value.Component.ActorTags.Contains(Tag))
+				{
+					Worker_EntityId Entity = NetDriver->PackageMap->GetEntityIdFromObject(Item.Key);
+					if (Entity != SpatialConstants::INVALID_ENTITY_ID)
+					{
+						AddEntityToWatch(Entity);
+					}
+				}
+			}
+		}
+	}
+}
+
+void USpatialNetDriverDebugContext::RemoveInterestOnTag(FName Tag)
+{
+	if (SemanticInterest.Remove(Tag) > 0)
+	{
+		CachedInterestSet.Empty();
+		bNeedToUpdateInterest = true;
+
+		TArray<Worker_EntityId_Key> EntityIds;
+		NetDriver->StaticComponentView->GetEntityIds(EntityIds);
+
+		for (auto Entity : EntityIds)
+		{
+			if (SpatialGDK::DebugComponent* DbgComp = NetDriver->StaticComponentView->GetComponentData<SpatialGDK::DebugComponent>(Entity))
+			{
+				if (!IsSetIntersectionEmpty(DbgComp->ActorTags, SemanticInterest))
+				{
+					AddEntityToWatch(Entity);
+				}
+			}
+		}
+
+		for (const auto& Item : ActorDebugInfo)
+		{
+			if (!Item.Value.bAdded || Item.Value.bUpdated)
+			{
+				if (!IsSetIntersectionEmpty(Item.Value.Component.ActorTags, SemanticInterest))
+				{
+					Worker_EntityId Entity = NetDriver->PackageMap->GetEntityIdFromObject(Item.Key);
+					if (Entity != SpatialConstants::INVALID_ENTITY_ID)
+					{
+						AddEntityToWatch(Entity);
+					}
+				}
+			}
+		}
+	}
+}
+
+void USpatialNetDriverDebugContext::KeepActorOnLocalWorker(AActor* Actor)
+{
+	if (Actor->HasAuthority())
+	{
+		DebugComponentView& Comp = GetDebugComponentView(Actor);
+		Comp.Component.DelegatedWorkerId = DebugStrategy->GetLocalVirtualWorkerId();
+		Comp.bUpdated = true;
+	}
+}
+
+void USpatialNetDriverDebugContext::DelegateTagToWorker(FName Tag, int32 WorkerId)
```

Comment:
Personal preference is to have two functions here, one for adding the delegation, and the other for removing.